### ISSUE TITLE
opt generates wrong code with "correlated-propagation,dce,flattencfg"

### ISSUE THREAD

**Issue by Bazoka13** [llvm/llvm-project#71068]
*******************************************************************************
OS and Platform:
Ubuntu 20.04.4 LTS
*******************************************************************************
clang version 17.0.4
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/ctc/llvm-releases/llvm-17.0.4/build/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/9
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/9
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Candidate multilib: x32;@mx32
Selected multilib: .;@m64
*******************************************************************************
Command Lines:
```shell
clang -O3 -mllvm -disable-llvm-optzns -c -emit-llvm a.c -o a.bc
opt -passes=correlated-propagation,dce,flattencfg a.bc -o a.opt.bc
clang a.opt.bc -o a.out
 ./a.out
Floating point exception (core dumped)
```
*******************************************************************************
Source Code:
```C
int a;
void b() {
  int c;
  unsigned int d = 6;
  c = 3 % (((a || 0) < 0) - d || 0);
}
int main() { b(); }
```

**Comment by dtcxzyw** [comment]
Sorry, I cannot reproduce this problem with the trunk version of LLVM. Maybe it has been fixed by some commits :)

**Comment by mahtohappy** [comment]
I was able to reproduce the error by building the LLVM trunk.

**Comment by dtcxzyw** [comment]
@mahtohappy @Bazoka13  Could you please provide the generated IR? (before/after opt)
```
clang -O3 -mllvm -disable-llvm-optzns -c -emit-llvm a.c -S -o a.ll
cat a.ll
opt -passes=correlated-propagation,dce,flattencfg a.ll -S -o a.opt.ll
cat a.opt.ll
```

**Comment by mahtohappy** [comment]
`; ModuleID = '71068.c'
source_filename = "71068.c"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@a = dso_local global i32 0, align 4

; Function Attrs: nounwind uwtable
define dso_local void @b() #0 {
entry:
  %c = alloca i32, align 4
  %d = alloca i32, align 4
  call void @llvm.lifetime.start.p0(i64 4, ptr %c) #2
  call void @llvm.lifetime.start.p0(i64 4, ptr %d) #2
  store i32 6, ptr %d, align 4, !tbaa !5
  %0 = load i32, ptr @a, align 4, !tbaa !5
  %tobool = icmp ne i32 %0, 0
  br i1 %tobool, label %lor.end, label %lor.rhs

lor.rhs:                                          ; preds = %entry
  br label %lor.end

lor.end:                                          ; preds = %lor.rhs, %entry
  %1 = phi i1 [ true, %entry ], [ false, %lor.rhs ]
  %lor.ext = zext i1 %1 to i32
  %cmp = icmp slt i32 %lor.ext, 0
  %conv = zext i1 %cmp to i32
  %2 = load i32, ptr %d, align 4, !tbaa !5
  %sub = sub i32 %conv, %2
  %tobool1 = icmp ne i32 %sub, 0
  br i1 %tobool1, label %lor.end3, label %lor.rhs2

lor.rhs2:                                         ; preds = %lor.end
  br label %lor.end3

lor.end3:                                         ; preds = %lor.rhs2, %lor.end
  %3 = phi i1 [ true, %lor.end ], [ false, %lor.rhs2 ]
  %lor.ext4 = zext i1 %3 to i32
  %rem = srem i32 3, %lor.ext4
  store i32 %rem, ptr %c, align 4, !tbaa !5
  call void @llvm.lifetime.end.p0(i64 4, ptr %d) #2
  call void @llvm.lifetime.end.p0(i64 4, ptr %c) #2
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite)
declare void @llvm.lifetime.start.p0(i64 immarg, ptr nocapture) #1

; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite)
declare void @llvm.lifetime.end.p0(i64 immarg, ptr nocapture) #1

; Function Attrs: nounwind uwtable
define dso_local i32 @main() #0 {
entry:
  call void @b()
  ret i32 0
}

attributes #0 = { nounwind uwtable "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cmov,+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
attributes #1 = { nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) }
attributes #2 = { nounwind }

!llvm.module.flags = !{!0, !1, !2, !3}
!llvm.ident = !{!4}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 8, !"PIC Level", i32 2}
!2 = !{i32 7, !"PIE Level", i32 2}
!3 = !{i32 7, !"uwtable", i32 2}
!4 = !{!"clang version 18.0.0 (https://github.com/mahtohappy/llvm-project.git 3b589fe6e68ce0b17515e4a88b5188abb8d6216d)"}
!5 = !{!6, !6, i64 0}
!6 = !{!"int", !7, i64 0}
!7 = !{!"omnipotent char", !8, i64 0}
!8 = !{!"Simple C/C++ TBAA"}`

After opt
`; ModuleID = 'a.ll'
source_filename = "71068.c"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@a = dso_local global i32 0, align 4

; Function Attrs: nounwind uwtable
define dso_local void @b() #0 {
entry:
  %c = alloca i32, align 4
  %d = alloca i32, align 4
  call void @llvm.lifetime.start.p0(i64 4, ptr nonnull %c) #2
  call void @llvm.lifetime.start.p0(i64 4, ptr nonnull %d) #2
  store i32 6, ptr %d, align 4, !tbaa !5
  %0 = load i32, ptr @a, align 4, !tbaa !5
  %tobool = icmp ne i32 %0, 0
  %conv = zext i1 false to i32
  %1 = load i32, ptr %d, align 4, !tbaa !5
  %sub = sub i32 %conv, %1
  %tobool1 = icmp ne i32 %sub, 0
  %2 = and i1 %tobool, %tobool1
  br i1 %2, label %lor.end3, label %lor.rhs2

lor.rhs2:                                         ; preds = %entry
  br label %lor.end3

lor.end3:                                         ; preds = %lor.rhs2, %entry
  %3 = phi i1 [ true, %entry ], [ false, %lor.rhs2 ]
  %lor.ext4 = zext i1 %3 to i32
  %rem1.rhs.trunc = trunc i32 %lor.ext4 to i8
  %rem12 = urem i8 3, %rem1.rhs.trunc
  %rem1.zext = zext i8 %rem12 to i32
  store i32 %rem1.zext, ptr %c, align 4, !tbaa !5
  call void @llvm.lifetime.end.p0(i64 4, ptr nonnull %d) #2
  call void @llvm.lifetime.end.p0(i64 4, ptr nonnull %c) #2
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite)
declare void @llvm.lifetime.start.p0(i64 immarg, ptr nocapture) #1

; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite)
declare void @llvm.lifetime.end.p0(i64 immarg, ptr nocapture) #1

; Function Attrs: nounwind uwtable
define dso_local i32 @main() #0 {
entry:
  call void @b()
  ret i32 0
}

attributes #0 = { nounwind uwtable "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cmov,+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
attributes #1 = { nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) }
attributes #2 = { nounwind }

!llvm.module.flags = !{!0, !1, !2, !3}
!llvm.ident = !{!4}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 8, !"PIC Level", i32 2}
!2 = !{i32 7, !"PIE Level", i32 2}
!3 = !{i32 7, !"uwtable", i32 2}
!4 = !{!"clang version 18.0.0 (https://github.com/mahtohappy/llvm-project.git 3b589fe6e68ce0b17515e4a88b5188abb8d6216d)"}
!5 = !{!6, !6, i64 0}
!6 = !{!"int", !7, i64 0}
!7 = !{!"omnipotent char", !8, i64 0}
!8 = !{!"Simple C/C++ TBAA"}`

**Comment by mahtohappy** [comment]
I think ```flattencfg``` pass is the one that's creating problem. Running 
``` opt -passes=correlated-propagation,dce a.ll -o a.opt.ll -S ```

**Comment by mahtohappy** [comment]
I did some digging here and the crash is due to divide by zero error which is not allowed. In the optimized case the ```c = 3 % (((a || 0) < 0) - d || 0); ``` the expression ``` (((a || 0) < 0) - d || 0) ``` is evaluating to be zero while it should be -1. I'll work on fixing it.

**Comment by dtcxzyw** [comment]
I still failed to reproduce this issue: https://godbolt.org/z/nc5vMWhhx

**Comment by roife** [comment]
I believe this issue has been fixed by #81987.

**Comment by fhahn** [comment]
@roife thanks, closing.

@mahtohappy  please let us know if you still are able to reproduce this
